code/modules/atmospherics/machinery/components/trinary_devices/filter.dm

```diff
@@ -119,43 +119,46 @@
 		if(!removed)
 			return
 		var/datum/gas_mixture/filtered_out = new
-		filtered_out.temperature = removed.temperature
+		filtered_out.set_temperature(removed.temperature())
 
 		switch(filter_type)
 			if(FILTER_TOXINS)
-				filtered_out.toxins = removed.toxins
-				removed.toxins = 0
+				filtered_out.set_toxins(removed.toxins())
+				removed.set_toxins(0)
 
-				filtered_out.agent_b = removed.agent_b
-				removed.agent_b = 0
+				filtered_out.set_agent_b(removed.agent_b())
+				removed.set_agent_b(0)
 
 			if(FILTER_OXYGEN)
-				filtered_out.oxygen = removed.oxygen
-				removed.oxygen = 0
+				filtered_out.set_oxygen(removed.oxygen())
+				removed.set_oxygen(0)
 
 			if(FILTER_NITROGEN)
-				filtered_out.nitrogen = removed.nitrogen
-				removed.nitrogen = 0
+				filtered_out.set_nitrogen(removed.nitrogen())
+				removed.set_nitrogen(0)
 
 			if(FILTER_CO2)
-				filtered_out.carbon_dioxide = removed.carbon_dioxide
-				removed.carbon_dioxide = 0
+				filtered_out.set_carbon_dioxide(removed.carbon_dioxide())
+				removed.set_carbon_dioxide(0)
 
 			if(FILTER_N2O)
-				filtered_out.sleeping_agent = removed.sleeping_agent
-				removed.sleeping_agent = 0
+				filtered_out.set_sleeping_agent(removed.sleeping_agent())
+				removed.set_sleeping_agent(0)
 			else
 				filtered_out = null
 
 
 		air2.merge(filtered_out)
 		air3.merge(removed)
 
-	parent2.update = 1
+	if(!QDELETED(parent1))
+		parent1.update = 1
 
-	parent3.update = 1
+	if(!QDELETED(parent2))
+		parent2.update = 1
 
-	parent1.update = 1
+	if(!QDELETED(parent3))
+		parent3.update = 1
```

Comment:
This fixes the `cannot modify null.update`, but also worries me from overhead.

That said we have gained so much tick budget from this I feel like this should be fine now.

Reply:
I'll be doing more performance passes on SSair later, so I'll take a look at this then. Seems to be fine in practice for now.